mfs/ops.go

```diff
@@ -3,10 +3,154 @@ package mfs
 import (
 	"errors"
 	"fmt"
+	"os"
+	gopath "path"
 	"strings"
+
+	dag "github.com/ipfs/go-ipfs/merkledag"
 )
 
-func rootLookup(r *Root, path string) (FSNode, error) {
+// Mv moves the file or directory at 'src' to 'dst'
+func Mv(r *Root, src, dst string) error {
+	srcDir, srcFname := gopath.Split(src)
+
+	var dstDirStr string
+	var filename string
+	if dst[len(dst)-1] == '/' {
+		dstDirStr = dst
+		filename = srcFname
+	} else {
+		dstDirStr, filename = gopath.Split(dst)
+	}
+
+	// get parent directories of both src and dest first
+	dstDir, err := lookupDir(r, dstDirStr)
+	if err != nil {
+		return err
+	}
+
+	srcDirObj, err := lookupDir(r, srcDir)
+	if err != nil {
+		return err
+	}
+
+	srcObj, err := srcDirObj.Child(srcFname)
+	if err != nil {
+		return err
+	}
+
+	nd, err := srcObj.GetNode()
+	if err != nil {
+		return err
+	}
+
+	fsn, err := dstDir.Child(filename)
+	if err == nil {
+		switch n := fsn.(type) {
+		case *File:
+			_ = dstDir.Unlink(filename)
+		case *Directory:
+			dstDir = n
+		default:
+			return fmt.Errorf("unexpected type at path: %s", dst)
+		}
+	} else if err != os.ErrNotExist {
+		return err
+	}
+
+	err = dstDir.AddChild(filename, nd)
+	if err != nil {
+		return err
+	}
+
+	err = srcDirObj.Unlink(srcFname)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func lookupDir(r *Root, path string) (*Directory, error) {
+	di, err := Lookup(r, path)
+	if err != nil {
+		return nil, err
+	}
+
+	d, ok := di.(*Directory)
+	if !ok {
+		return nil, fmt.Errorf("%s is not a directory", path)
+	}
+
+	return d, nil
+}
+
+// PutNode inserts 'nd' at 'path' in the given mfs
+func PutNode(r *Root, path string, nd *dag.Node) error {
+	dirp, filename := gopath.Split(path)
+
+	parent, err := Lookup(r, dirp)
+	if err != nil {
+		return fmt.Errorf("lookup '%s' failed: %s", dirp, err)
+	}
+
+	pdir, ok := parent.(*Directory)
+	if !ok {
+		return fmt.Errorf("%s is not a directory", dirp)
+	}
+
+	return pdir.AddChild(filename, nd)
+}
+
+// Mkdir creates a directory at 'path' under the directory 'd', creating
+// intermediary directories as needed if 'parents' is set to true
+func Mkdir(r *Root, path string, parents bool) error {
+	parts := strings.Split(path, "/")
+	if parts[0] == "" {
```

Comment:
Hrm, maybe i should error out if it doesnt begin with a '/'.

In the commands lib code, it verifies that paths begin with a slash.